<launch>

    <!-- mono or stereo and what ros bag to play -->
    <arg name="max_cameras" default="2" />
    <arg name="use_stereo"  default="true" />
    <arg name="bag_start"   default="0" /> <!-- v1-2: 0, mh1: 40, mh2: 35, mh3-5: 20 -->
<!--     <arg name="bag"         default="/home/zzq/DataSet/euroc_mav/V1_01_easy.bag" /> -->
      <!-- we need to change function load_posegraphs2, get_multiKF_feature, get_loopkfs, get_feature_jacobian_kf3 from Euroc to YQ-->
    <!-- imu starting thresholds -->
    <arg name="init_window_time"  default="0.5" />     <!--0.75  1-->  
    <arg name="init_imu_thresh"   default="0.3" />    <!--2   1-->

    <!-- saving trajectory path and timing information -->
    <arg name="dosave"      default="true" />
    <arg name="dotime"      default="true" />
    <arg name="path_est"    default="/tmp/traj_estimate.txt" />
    <arg name="path_rect_est"    default="/tmp/traj_rect_estimate.txt" />
    <arg name="path_time"   default="/tmp/traj_timing.txt" />
    <!-- <arg name="points_filename"    default="/tmp/match_points.txt" /> -->


    <!-- MASTER NODE! -->
    <node name="run_subscribe_msckf" pkg="ov_msckf" type="run_subscribe_msckf" output="screen" clear_params="true" required="true">

        <!-- bag topics -->
        <param name="topic_imu"      type="string" value="/xsens/imu_data" />
         <param name="topic_camera0"  type="string" value="/stereo_grey/left/image_raw" /> 
        <param name="topic_camera1"  type="string" value="/stereo_grey/right/image_raw" /> 
<!--         <param name="topic_camera0"  type="string" value="/pointgrey/left/image_raw/compressed" />-->
 <!--        <param name="topic_camera1"  type="string" value="/pointgrey/right/image_raw/compressed" />-->

        <param name="use_prior_map" type="bool" value="false" />
        <param name="map_save_path" type="string" value="/home/zzq/Code/map_open_vins_ros/src/open_vins/matches/" />
        <!-- <param name="pose_graph_filename" type="string" value="20170827_loc_timeq_timer_2dq_2dr_3dr.txt" /> -->
        <param name="pose_graph_filename" type="string" value="20170828_loc_timeq_timer_2dq_2dr_3dr.txt" />
        <!-- <param name="pose_graph_filename" type="string" value="20180129_loc_timeq_timer_2dq_2dr_3dr.txt" /> -->
        <!-- <param name="path_gt"     type="string" value="/home/zzq/Code/map_open_vins_ros/src/open_vins/matches/gt_imu_20170828.txt" /> -->
        <param name="keyframe_pose_filename" type="string" value="gt_cam_20170823.txt" />
        <param name="points_filename" type="string" value="/tmp/match_points.txt" />
        <param name="save_match_points" type="bool" value="true" />
        <param name="transform_filename" type="string" value="/tmp/transform.txt" />
        <param name="save_transform" type="bool" value="true" />
        <param name="match_num_thred" type="int" value="20" />
        <param name="pixel_error" type="double" value="3.0" />
        <param name="transform_cov" type="double" value="20.0" />
        <param name="keyframe_cov" type="double" value="0.1"/>

        

        <param name="brief_pattern_filename" type="string" value="brief_pattern.yml" />
        <param name="voc_filename" type="string" value="brief_k10L6.bin" />

        <param name="kf_match_interval" type="double" value="10" />
        <param name="use_robust_match" type="bool" value="true" />

        <!-- world/filter parameters -->
        <param name="use_fej"                type="bool"   value="true" />
        <param name="use_imuavg"             type="bool"   value="true" />
        <param name="use_rk4int"             type="bool"   value="true" />
        <param name="use_stereo"             type="bool"   value="$(arg use_stereo)" />
        <param name="calib_cam_extrinsics"   type="bool"   value="true" />
        <param name="calib_cam_intrinsics"   type="bool"   value="true" />
        <param name="calib_cam_timeoffset"   type="bool"   value="true" />    
        <param name="calib_camimu_dt"        type="double" value="0.0" />
        <param name="max_clones"             type="int"    value="15" />
        <param name="max_slam"               type="int"    value="50" /> <!--75-->
        <param name="max_slam_in_update"     type="int"    value="25" /> <!-- 25 seems to work well -->
        <param name="max_msckf_in_update"    type="int"    value="999" />
        <param name="max_cameras"            type="int"    value="$(arg max_cameras)" />
        <param name="dt_slam_delay"          type="double" value="3" />
        <param name="init_window_time"       type="double" value="$(arg init_window_time)" />
        <param name="init_imu_thresh"        type="double" value="$(arg init_imu_thresh)" />
        <rosparam param="gravity">[0.0,0.0,9.81]</rosparam>
        <param name="feat_rep_msckf"         type="string" value="GLOBAL_3D" />
        <param name="feat_rep_slam"          type="string" value="ANCHORED_FULL_INVERSE_DEPTH" />
        <param name="feat_rep_aruco"         type="string" value="ANCHORED_FULL_INVERSE_DEPTH" />

        <!-- zero velocity update parameters -->
        <param name="try_zupt"               type="bool"   value="false" />
        <param name="zupt_chi2_multipler"    type="int"    value="1" />
        <param name="zupt_max_velocity"      type="double" value="1" />
        <param name="zupt_noise_multiplier"  type="double" value="50" />

        <!-- timing statistics recording -->
        <param name="record_timing_information"   type="bool"   value="$(arg dotime)" />
        <param name="record_timing_filepath"      type="string" value="$(arg path_time)" />

        <!-- tracker/extractor properties -->
        <param name="use_klt"            type="bool"   value="true" />
        <param name="num_pts"            type="int"    value="200" />
        <param name="fast_threshold"     type="int"    value="10" />
        <param name="grid_x"             type="int"    value="10" />
        <param name="grid_y"             type="int"    value="5" />
        <param name="min_px_dist"        type="int"    value="10" />
        <param name="knn_ratio"          type="double" value="0.70" />
        <param name="downsample_cameras" type="bool"   value="false" />

        <!-- aruco tag/mapping properties -->
        <param name="use_aruco"        type="bool"   value="false" />
        <param name="num_aruco"        type="int"    value="1024" />
        <param name="downsize_aruco"   type="bool"   value="true" />

        <!-- sensor noise values / update -->
        <param name="up_msckf_sigma_px"            type="double"   value="1" />
        <param name="up_msckf_chi2_multipler"      type="double"   value="1" />
        <param name="up_slam_sigma_px"             type="double"   value="1" />
        <param name="up_slam_chi2_multipler"       type="double"   value="1" />
        <param name="up_aruco_sigma_px"            type="double"   value="1" />
        <param name="up_aruco_chi2_multipler"      type="double"   value="1" />
        <param name="gyroscope_noise_density"      type="double"   value="8.49e-2" />
        <param name="gyroscope_random_walk"        type="double"   value="1.26e-3" />
        <param name="accelerometer_noise_density"  type="double"   value="1.82e-1" />
        <param name="accelerometer_random_walk"    type="double"   value="1.9e-3" />

        <!-- camera intrinsics -->
        <rosparam param="cam0_wh">[1440, 1080]</rosparam>   <!-- rosparam以yaml来读数据，通常用于矩阵的读取-->
        <rosparam param="cam1_wh">[1440, 1080]</rosparam>
        <param name="cam0_is_fisheye" type="bool" value="false" />
        <param name="cam1_is_fisheye" type="bool" value="false" />
        <rosparam param="cam0_k">[1193.4,1193.8,746.6,516.7]</rosparam>
        <!-- <rosparam param="cam0_d">[-0.09476847846825949, 0.07976950266449231, -0.001130892244965163, -0.002061147569213533]</rosparam> -->
        <rosparam param="cam0_d">[-0.0791, 0.0976, 0, 0]</rosparam>
        <rosparam param="cam1_k">[1192.7, 1192.7, 783.6, 557]</rosparam>
        <!-- <rosparam param="cam1_d">[-0.09662792944244207, 0.08660458061962732, -0.003583726041952065, 8.94486463654094e-05]</rosparam> -->
        <rosparam param="cam1_d">[-0.0765, 0.0832, 0, 0]</rosparam>
        <!-- camera extrinsics -->
        <rosparam param="T_C0toI">
            [
            -0.0077, -0.0039, 1.0000, 0.035,
               -0.9999, -0.0065, -0.0077, -0.025,
                0.0066, -1.000, -0.0038, 0.0012,
                     0, 0, 0, 1.0000
            ]
        </rosparam>
         
          <!-- <rosparam param="T_C1toI">
            [
                   0.0202174, -0.0210638,   0.999557,   0.154156,
                     -1.0042, -0.0185661,  0.0197384,  0.0798734,
                    0.0180645,  -0.999678, -0.0215781, -0.0717355,
                    0  ,        0  ,        0 ,         1

            ]
        </rosparam> -->
        <rosparam param="T_C1toI">
            [
                   0.001500555959000,  -0.001516710184000,    1.000035170085000,    0.0406,
                      -0.999939509041000,  -0.004557318517000,   0.001518076031000,    -0.5230,
                   0.004625530036000,   -1.000016942174000,  -0.001474439724000,  0.003407343440000,
                             0,           0,           0,           1

            ]
        </rosparam>


    </node>
    <node type="rviz" name="rviz" pkg="rviz" args="-d $(find ov_msckf)/launch/display_4.rviz" />

    <!-- play the dataset -->
<!--     <node pkg="rosbag" type="play" name="rosbag" args="-d 1 -s $(arg bag_start) $(arg bag)" required="true"/> -->

    <!-- record the trajectory if enabled -->
    <group if="$(arg dosave)">
        <node name="recorder_estimate" pkg="ov_eval" type="pose_to_file" output="screen" required="true">
            <param name="topic"      type="str" value="/ov_msckf/poseimu" />
            <param name="topic_type" type="str" value="PoseWithCovarianceStamped" />
            <param name="output"     type="str" value="$(arg path_est)" />
            <param name="topic_rect"      type="str" value="/ov_msckf/poseimu_rect" />
            <param name="topic_type_rect" type="str" value="PoseWithCovarianceStamped" />
            <param name="output_rect"     type="str" value="$(arg path_rect_est)" />
        </node>
    </group>


</launch>